pkg/bootflow/types/state.go

```diff
@@ -0,0 +1,125 @@
+package types
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+)
+
+// State describes a virtual state of a machine being booted
+type State struct {
+	// Do not mutate these values from the outside, the mutability
+	// is owned by State itself.
+
+	SystemArtifacts SystemArtifacts
+	TrustChains     TrustChains
+	CurrentFlow     Flow
+	CurrentStepIdx  uint
+	VerifiedData    []VerifiedData
+}
+
+func typeMapKey(i interface{}) reflect.Type {
+	k := reflect.TypeOf(i)
+	for k.Kind() == reflect.Ptr {
+		k = k.Elem()
+	}
+	return k
+}
+
+func NewState() *State {
+	return &State{
+		SystemArtifacts: map[reflect.Type]SystemArtifact{},
+		TrustChains:     map[reflect.Type]TrustChain{},
+	}
+}
+
+func (state *State) SetFlow(flow Flow, stepIdx uint) {
```

Comment:
Does the state need to be rolled back when transitioning to new flow? I think we agree we it's hard to guarantee it's done correctly. Should the boot process restart from a clean state (while keeping track of the encountered issues), when a `Flow` fails and we need to switch? Heavy operations could be cached.

Reply:
> Does the state need to be rolled back when transitioning to new flow?

In the current proposal we do not roll back.

> Should the boot process restart from a clean state (while keeping track of the encountered issues), when a `Flow` fails and we need to switch?

In the current proposal we are not restarting the boot process.

---

The `SetFlow` is like `JMP`, not like `CALL`. There is no returning to the original `Flow`, no saving state. Just sequentially execute steps until there will be no more steps to execute.

Basically a `Flow` might be of the of, for example (in pseudo-code):
```
Root {
    IfFITPresent{SetFlow{Intel}}
    IfAMDManifestPresent{SetFlow{AMD}}
}

Intel {
    IfBPMPresent{SetFlow{CBnT}}
    SetFlow{LegacyTXT}
}

CBnT {
    ValidateACM{Failure: SetFlow{CBnTFailure}}
    ValidateKM{Failure: SetFlow{CBnTFailure}}
    ValidateBPM{Failure: SetFlow{CBnTFailure}}
    ValidateIBB{Failure: SetFlow{CBnTFailure}}
    TPMInit(3)
    MeasurePCR0DATA
    SetFlow{IntelResetVector}
}

CBnTFailure {
    IfFITAddrIs{specialFitType, specialAddrValue, CBnT0TFailureMeta} // check if a special FIT entry has a special value
    SetFlow{IntelResetVector}
}

CBnT0TFailureMeta {
    MeasurePCR0DATAFailure
    SetFlow{IntelResetVector}
}

IntelResetVector {
    IfOCP{SetFlow{OCPPEI}}
    IfHPE{SetFlow{HPEPEI}}
    IfLenovo{SetFlow{LenovoPEI}}
}

OCPPEI {
    TPMInitLazy(3)
    MeasureUEFIGUID(0, OCP_PEI_GUID)
    MeasureSeparator(0)
    SetFlow{OCPDXE}
}

OCPDXE {
    MeasurePCDVariable(0, "FirmwareVendorVersion")
    MeasureUEFIGUID(0, OCP_DXE_GUID)
    MeasureSeparator
}
```

Reply:
Ok, as long as the contract is clear that a flow should be built to leave consistent state behind. Hard to enforce, but this is what should happen in real world. It's not always the case, but this doesn't mean we should do it right here :) 

As an example, Intel ACM I think behaves slightly differently (e.g. requirement to cap PCR registers after initializing TPM if NVRAM indexes are provisioned in wrong way). Intel ACM would actually have:

```
    TPMInit(3)
    ValidateACM{Failure: SetFlow{CBnTFailure}}
    ValidateKM{Failure: SetFlow{CBnTFailure}}
    ValidateBPM{Failure: SetFlow{CBnTFailure}}
    ValidateIBB{Failure: SetFlow{CBnTFailure}}
    MeasurePCR0DATA
    SetFlow{IntelResetVector}
```

But ok, Intel ACM is the broken one, no need to replicate broken behavior :)

Reply:
> Intel ACM would actually have:

I'm actually not sure. It actually might be than `TPMInit(3)` happens only on on-need basis (not earlier).

Reply:
Hm... that looks counter intitive. So different flows should know about each other + that redefinition of the flow for the step.

I would propose to make Flow an interface that returns the next step, this will free flows to know about each other, only share the steps (which they are doing anyway)

```
  type Flow []Step
  
  type Flow interface {
       Next() (Step, error) // returns nil if this is the last step or error is smth went wrong and the next step could not be determined
  }
```

Reply:
> Ok, as long as the contract is clear that a flow should be built to leave consistent state behind. 

One of the major idea of this new design: to be close to how actual boot process works (to be able to reproduce its behavior). And unfortunately in the actual boot process we have no such guarantees, thus in these packages not having this guarantees is considered a feature (not a bug): if something in a real boot leaves an unclean state, we want to have exactly the same thing here.

> So different flows should know about each other + that redefinition of the flow for the step.

In reality (in a real boot process) they do know about entry points of each other.

> I would propose to make Flow an interface that returns the next step, this will free flows to know about each other, only share the steps (which they are doing anyway)

Sorry, I do not understand. They will still to know about each other to be able to jump to each other. Otherwise we need to denormalize the pseudo-code tree given above (and make it a bunch of flat sequential steps instead of being a tree).